aemcmc/gibbs.py

```diff
@@ -364,6 +423,129 @@ def nbinom_horseshoe_step(
+
+    Returns
+    -------
+    (outputs, updates): tuple
+        A symbolic description of the sampling result to be used to
+        compile a sampling function.
+
+    Notes
+    -----
+    The ``z`` expression in section 2.2 of [1] seems to
+    omit division by the Polya-Gamma auxilliary variables whereas [2] and [3]
+    explicitely include it. We found that including the division results in
+    accurate posterior samples for the regression coefficients. It is also
+    worth noting that the :math:`\sigma^2` parameter is not sampled directly
+    in the negative binomial regression problem and thus set to 1 [2].
+
+    References
+    ----------
+    ..[1] Makalic, Enes & Schmidt, Daniel. (2015). A Simple Sampler for the
+          Horseshoe Estimator. 10.1109/LSP.2015.2503725.
+    ..[2] Makalic, Enes & Schmidt, Daniel. (2016). High-Dimensional Bayesian
+          Regularised Regression with the BayesReg Package.
+    ..[3] Neelon, Brian. (2019). Bayesian Zero-Inflated Negative Binomial
+          Regression Based on Pólya-Gamma Mixtures. Bayesian Anal.
+          2019 September ; 14(3): 829–855. doi:10.1214/18-ba1132.
+
+    """
+    T = y.shape[0].eval()
+
+    def nbinom_horseshoe_step(
+        beta: TensorVariable,
+        lmbda: TensorVariable,
+        tau: TensorVariable,
+        r: TensorVariable,
+        phi: TensorVariable,
+        y: TensorVariable,
+        X: TensorVariable,
+    ) -> Tuple[
+        TensorVariable, TensorVariable, TensorVariable, TensorVariable, TensorVariable
+    ]:
+        """Complete one full update of the gibbs sampler and return the new state
+        of the posterior conditional parameters.
+
+        Parameters
+        ----------
+        beta: Tensorvariable
+            Coefficients (other than intercept) of the regression model.
+        lmbda
+            Inverse of the local shrinkage parameter of the horseshoe prior.
+        tau
+            Inverse of the global shrinkage parameters of the horseshoe prior.
+        y: TensorVariable
+            The observed count data.
+        X: TensorVariable
+            The covariate matrix.
+        r: TensorVariable
+            The "number of successes" parameter of the negative binomial disribution
+            used to model the data.
+
+        """
+        xb = X @ beta
+        # y_r = at.switch(y + r < 0.001,0.001, y + r )
+        w = srng.gen(polyagamma, y + r, xb)
+        z = 0.5 * (y - r) / w
+
+        lmbda_inv = 1.0 / lmbda
+        tau_inv = 1.0 / tau
+        beta_new = update_beta(srng, w, lmbda_inv * tau_inv, X, z)
+
+        lmbda_inv_new, tau_inv_new = horseshoe_step(
+            srng, beta_new, 1.0, lmbda_inv, tau_inv
+        )
+
+        r_new, phi_new = dispersion_term_step(srng, phi, r, w, y, xb, T)
+
+        return beta_new, 1.0 / lmbda_inv_new, 1.0 / tau_inv_new, r_new, phi_new
+
+    X, beta_rv, lmbda_rv, tau_rv, r_rv = nbinom_horseshoe_match_with_dispersion(Y_rv)
+
+    phi_rv = srng.gamma(1, 1, size=1)
```

Comment:
What are these `phi*` terms?  Also, a sampling step for $r$ should only require the previous sample value of $r$ and the $\pi_i$ that make up $y_i \sim \operatorname{NB}\left(r, \pi_i\right)$.